File: p3/p34337Tests/TestC.java
import java.lang.reflect.InvocationTargetException;
import java.util.List;

/**
 * Test:  Verify servers will wait if the
 * buffet is too full.
 * 
 * @version 1.0
 */
public class TestC {
  /**
   * Scot gets there first, with 9 slices
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable scot(final Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        Util.sleep(1);
        Util.printAdd("Scot", 9, SliceType.Meat, buffet.AddPizza(9, SliceType.Meat));
      }
    };
  }

  /**
   * Shirlee gets there next, but can't drop off all her slices
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable shirlee(final Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        Util.sleep(2);
        Util.printAdd("Shirlee", 12, SliceType.Works, buffet.AddPizza(12, SliceType.Works));
      }
    };
  }

  /**
   * Marnie takes some slices after three seconds.  This will let shirlee finish
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable marnie(final Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        Util.sleep(3);
        Util.printSlices("Marnie", buffet.TakeAny(2));
      }
    };
  }

  /**
   * Create, run, and join actor threads
   */
  public static void main(String[] args) throws NumberFormatException, NoSuchMethodException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
    if (args.length != 1) {
      throw new IllegalArgumentException("Parameter(s): <Buffet Class>");
    }
    // Initialize the buffet our threads are using
    Buffet buffet = Util.getBuffet(args[0], 20);

    // Make a few server threads
    Util.run(List.of(new Thread(marnie(buffet)),
        new Thread(shirlee(buffet)), new Thread(scot(buffet))));
  }
}

File: p3/p5test-1/testc.cpp
/**
 * Simple test case for the monitor.  A server will have to wait if the
 * buffet is too full
 */
#include <iostream>
#include <stdexcept>
#include <vector>
#include <unistd.h>
#include <pthread.h>

#include "buffetmutex.h"
#include "util.h"

using namespace std;

/** Scot gets there first, with 9 slices. */
void *scot(void *arg) {
  sleep(1);
  Buffet *buffet = (Buffet*) arg;
  if (buffet->AddPizza(9, Meat)) {
    cout << ts() << ": Scot dropped off 9 slices" << endl;
  } else {
    cerr << ts() << ": Scot couldn't drop off everything" << endl;
  }

  return NULL;
}

/** Shirlee gets there next, but can't drop off all her slices. */
void *shirlee(void *arg) {
  sleep(2);
  Buffet *buffet = (Buffet*) arg;
  if (buffet->AddPizza(12, Works)) {
    cout << ts() << ": Shirlee dropped off 12 slices" << endl;
  } else {
    cerr << ts() << ": Shirlee couldn't drop off everything" << endl;
  }

  return NULL;
}

/** Marnie takes some slices after three seconds.  This will let shirlee finish. */
void *marnie(void *arg) {
  sleep(3);
  Buffet *buffet = (Buffet*) arg;
  printSlices("Marnie", buffet->TakeAny(2));

  return NULL;
}

/** Create, run, and join actor threads */
int main() {
  // Initialize the monitor our threads are using
  Buffet buffet(20);

  // Make a few agent threads
  const int NOTHREADS = 3;
  pthread_t thread[NOTHREADS];

  if (pthread_create(thread + 0, NULL, marnie, &buffet) != 0 ||
      pthread_create(thread + 1, NULL, shirlee, &buffet) != 0 ||
      pthread_create(thread + 2, NULL, scot, &buffet) != 0) {
    fail("Can't make a thread we need");
  }

  // When we're done, join with all the threads
  for (int i=0; i < NOTHREADS; i++) {
    pthread_join(thread[i], NULL);
  }

  return 0;
}
